lib/ansible/modules/cloud/digital_ocean/digital_ocean_tag.py

```diff
@@ -189,45 +189,65 @@ def core(module):
+    if response.status_code == 401:
+        module.fail_json(msg='Failed to login using api_token, please verify '
+                             'validity of api_token')
+    if state == 'present':
+        response = rest.get('tags/{0}'.format(name))
         status_code = response.status_code
-        json = response.json
-        if status_code == 201:
-            changed = True
-        elif status_code == 422:
+        resp_json = response.json
+        changed = False
+        if status_code == 200 and resp_json['tag']['name'] == name:
             changed = False
         else:
-            module.exit_json(changed=False, data=json)
+            # Ensure Tag exists
+            response = rest.post("tags", data={'name': name})
+            status_code = response.status_code
+            resp_json = response.json
+            if status_code == 201:
+                changed = True
+            elif status_code == 422:
+                changed = False
+            else:
+                module.exit_json(changed=False, data=resp_json)
 
         if resource_id is None:
             # No resource defined, we're done.
-            if json is None:
-                module.exit_json(changed=changed, data=json)
-            else:
-                module.exit_json(changed=changed, data=json)
+            module.exit_json(changed=changed, data=resp_json)
         else:
-            # Tag a resource
-            url = "tags/{}/resources".format(name)
-            payload = {
-                'resources': [{
-                    'resource_id': resource_id,
-                    'resource_type': resource_type}]}
-            response = rest.post(url, data=payload)
-            if response.status_code == 204:
-                module.exit_json(changed=True)
+            # Check if resource is already tagged or not
+            found = False
+            url = "{0}?tag_name={1}".format(resource_type, name)
+            response = rest.get(url)
+            status_code = response.status_code
+            resp_json = response.json
+            if status_code == 200:
+                for resource in resp_json['droplets']:
+                    if not found and resource['id'] == int(resource_id):
+                        found = True
+                        break
+                if not found:
+                    # If resource is not tagged, tag a resource
+                    url = "tags/{}/resources".format(name)
+                    payload = {
+                        'resources': [{
+                            'resource_id': resource_id,
+                            'resource_type': resource_type}]}
+                    response = rest.post(url, data=payload)
+                    if response.status_code == 204:
+                        module.exit_json(changed=True)
+                    else:
+                        module.fail_json(msg="error tagging resource '{}': {}".format(resource_id, response.json["message"]))
```

Comment:
IMHO this is not python 2.6 compatible (https://docs.python.org/2/library/string.html#format-string-syntax). I usually avoid format at all and use the % syntax.